### PR TITLE
[WIP] fixes beat timings

### PR DESCRIPTION
**Pull request by bmcfee**
See #36.

Possibly fixes some synthesis alignment issues as well.

### CODE REVIEW

**Comment by bmcfee** [comment]
Could someone (ie @tkell ) take a look at this?

I fixed the quantization bug and a minor hiccup in the sample placement, but i think it's still behaving incorrectly.  If I run reverse.py on a full-length track, it seems to put the entire second half of the track up front in forward-order, and then scatters reversed beats at the end.  Not sure what exactly is going on, and don't have time to dig into it more right now.

**Comment by tkell** [comment]
Will do some wrestling, you bet.

**Comment by tkell** [comment]
@bmcfee, looks like `beats` has 11 beats in it, and the last beat is 3 seconds long.  I'll do some math.

**Comment by tkell** [comment]
So, when using the example file, which is 16 beats long:  

`tempo, beat_frames = librosa.beat.beat_track(y=y_mono, sr=self.analysis_sample_rate, trim=False)` returns a 10 element array - which, I think, is why we have a long last beat.  Is this a librosa thing?

**Comment by bmcfee** [comment]
> returns a 10 element array - which, I think, is why we have a long last beat. Is this a librosa thing?

It can be, if `trim=True` and the there isn't a strong onset to grab at the end.  I don't think that should be the case here though.

**Comment by tkell** [comment]
K, let me try some things with trim = false...

**Comment by bmcfee** [comment]
oooh, i think i see what's going on.  Why are there two sample rates here? https://github.com/algorithmic-music-exploration/amen/blob/master/amen/audio.py#L34

**Comment by bmcfee** [comment]
nm, latest push should fix that.  My bad for not paying attention.

File: amen/audio.py [hunk 1]
```diff
@@ -108,10 +108,13 @@ def _get_beats(self):
         tempo, beat_frames = librosa.beat.beat_track(
             y=y_mono, sr=self.analysis_sample_rate, trim=False)
 
-        # convert frames to times
-        beat_times = librosa.frames_to_time(beat_frames, sr=self.sample_rate)
```
Comment (bmcfee) [inline amen/audio.py]: This line turned out to be the culprit.  I think this was only obvious when the track was long enough for the difference between sample_rate and analysis_sample_rate to be observable.  +1 for realistic test data...

**Comment by tkell** [comment]
Gave this a test - looks good to me, though I suspect that some nosetests might need a tweak.

**Comment by bmcfee** [comment]
> nosetests might need a tweak.

done.

File: amen/synthesize.py [hunk 2]
```diff
@@ -80,17 +80,17 @@ def synthesize(inputs):
...
         left_start = starting_sample + left_offsets[0] + initial_offset
-        left_end = ending_sample + left_offsets[1] + initial_offset
         right_start = starting_sample + right_offsets[0] + initial_offset
-        right_end = ending_sample + right_offsets[1] + initial_offset
 
         # add the data from each channel to the array
-        sparse_array[0, left_start:left_end] += resampled_audio[0]
-        sparse_array[1, right_start:right_end] += resampled_audio[1]
+        sparse_array[0, left_start:left_start+len(resampled_audio[0])] += resampled_audio[0]
+        sparse_array[1, right_start:right_start+len(resampled_audio[1])] += resampled_audio[1]
```
Comment (tkell) [inline amen/synthesize.py]: Remind me to refactor this so we don't get have to do `left_offsets[0]` and can just do `left_offset`.

**Comment by tkell** [comment]
:shipit:

(Better get back to 97% coverage)

**Comment by tkell** [comment]
Refactoring done, things passing, think this is good.

**Comment by bmcfee** [comment]
Eh, coverage is only down because i removed dead (but covered) code.  The uncovered bit is just a repr string, so i'm not worried.

@tkell merge away?

**Pull request merged**